scikit-hep/pyhf
# fix: Report expected p-values based on their quantiles under background hypotheses

# Description

* Resolves #1161 
* Resolves #1264

Fixes issue reported by @kpachal

This changes the way we report expected limits

we used to
* find quantiles in test stat distribution [q1,q2,....q5]
* report CLs/p-values for those test stats [p1,p2,...p5]

however this can lead to orderings in which the p-values are not monotonic as the relationship between CLs and test statistic value is not necessarily monotonic..

we now
* compute a sample of Cls values based on test statistics sampled from toys
* report exected CLs/p-values  as quantiles of that those test stats [p1,p2,...p5]

This is also what ROOT does. It might present some future questions regarding #966 but is ok for now.

```
import pyhf
import json
import numpy as np
import matplotlib.pyplot as plt
import pyhf.contrib.viz.brazil as brazil

w = pyhf.Workspace(json.loads('''
{
    "channels": [
        {
            "name": "SR_combined",
            "samples": [
                {
                    "data": [4],
                    "modifiers": [
                        {"data": null,"name": "lumi","type": "lumi"},
                        {"data": null,"name": "mu","type": "normfactor"}
                    ],
                    "name": "signal"
                },
                {
                    "data": [0.1],
                    "modifiers": [
                        {"data": null,"name": "lumi","type": "lumi"},
                        {"data": {"hi": 2.0,"lo": 0.5},"name": "uncorr_bkguncrt","type": "normsys"}
                    ],
                    "name": "background"
                }
            ]
        }
    ],
    "measurements": [
        {
            "config": {
                "parameters": [
                    {"auxdata": [1.0],"bounds": [[0.5,1.5]],"inits": [1.0],"name": "lumi","sigmas": [0.01]},
                    {"bounds": [[0.0,10.0]],"inits": [1.0],"name": "mu"}
                ],
                "poi": "mu"
            },
            "name": "Measurement"
        }
    ],
    "observations": [{"data": [0],"name": "SR_combined"}
    ],
    "version": "1.0.0"
}
'''))
m = w.model()
d = w.data(m)

mutests = np.linspace(0.5,1.0,4)
results = [pyhf.infer.hypotest(mu,d,m,return_expected_set=True, calctype='toybased', ntoys = 15000) for mu in mutests]

f,ax = plt.subplots()
brazil.plot_results(ax,mutests,results)
plt.semilogy()
plt.ylim(1e-3,1)
plt.xlim(0,1.35)
```

produces this plot

![image](https://user-images.githubusercontent.com/2318083/98489098-d3aa7080-222c-11eb-88aa-b0fb8162a300.png)


ReadTheDocs build: https://pyhf.readthedocs.io/en/fix_for_kate/api.html#inference

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add two ROOT validation scripts for toys
* Use same initialization parameters for b and s+b toys
* Leave (expected) p-value calculation to calculators
   - Add pvalue and expected_pvalues to calculator API
* Toy-based calculator returns expected values based on CLs distribution quantiles

Co-authored-by: Matthew Feickert <matthew.feickert@cern.ch>
Co-authored-by: Giordon Stark <kratsg@gmail.com>
```
